fri/src/verifier.rs

```diff
@@ -0,0 +1,282 @@
+
+            if i == 0 {
+                if self.params.fri_step_list[0] != 0 {
+                    self.first_eval_point = Some(self.channel.draw_felem());
+                }
+            } else {
+                self.eval_points[i - 1] = self.channel.draw_felem();
+            }
+
+            if i < self.n_layers - 1 {
+                let coset_size = 1 << self.params.fri_step_list[i + 1];
+                let n_rows = self.params.fft_domains[basis_index].size() / coset_size;
+                let n_columns = coset_size;
+                let size_of_row = ((F::MODULUS_BIT_SIZE.div_ceil(8) * 8) as usize) * n_columns;
+
+                let commitment_scheme = make_commitment_scheme_verifier(
+                    size_of_row,
+                    n_rows,
+                    0,
+                    self.commitment_hashes.clone(),
+                    n_columns,
+                );
+
+                let mut table_verifier = TableVerifier::new(n_columns, commitment_scheme);
+                let _ = table_verifier.read_commitment(&mut self.channel);
+                self.table_verifiers.push(table_verifier);
+            }
+        }
+
+        if let Some(eval_points_test) = self.eval_points_test.as_ref() {
+            self.first_eval_point = Some(eval_points_test[0]);
+            self.eval_points[..(self.n_layers - 1)]
+                .copy_from_slice(&eval_points_test[1..((self.n_layers - 1) + 1)]);
+        }
+        self.read_last_layer_coefficients()?;
+        Ok(())
+    }
+
+    pub fn read_last_layer_coefficients(&mut self) -> Result<(), Box<dyn Error>> {
+        let fri_step_sum: usize = self.params.fri_step_list.iter().sum();
+        let last_layer_size = self.params.fft_domains[fri_step_sum].size();
+
+        let mut last_layer_coefficients_vector = if let Some(last_layer_coefficients_test) =
+            self.last_layer_coefficients_test.as_ref()
+        {
+            last_layer_coefficients_test.clone()
+        } else {
+            self.channel
+                .recv_felts(self.params.last_layer_degree_bound as usize)
+                .unwrap()
+        };
+
+        // pad last_layer_coefficients_vector with zeros to the size of last_layer_size
+        while last_layer_coefficients_vector.len() < last_layer_size {
+            last_layer_coefficients_vector.push(F::zero());
+        }
+
+        assert!(
+            self.params.last_layer_degree_bound as usize <= last_layer_size,
+            "last_layer_degree_bound ({}) must be <= last_layer_size ({})",
+            self.params.last_layer_degree_bound,
+            last_layer_size
+        );
+
+        let last_layer_basis_index = fri_step_sum;
+        let lde_domain = self.params.fft_domains[last_layer_basis_index];
+        let mut lde = MultiplicativeLDE::new(lde_domain, true);
+
+        lde.add_coeff(&last_layer_coefficients_vector);
+        let evals = lde.eval(lde_domain.element(0));
+        self.expected_last_layer = Some(evals[0].clone());
+
```

Comment:
Wait, the query point of the last layer is always g^0 ?

Reply:
Nope. `lde.eval()` was a batch evaluation where `lde` struct can have multiple polynomials. So `evals[0]` just means a first evaluation. not evaluation from g^0.